RNs-4.9.31 Added with Bug Fix Update (#45254)
openshift/openshift-docs: release_notes/ocp-4-9-release-notes.adoc

==== Updating

To upgrade an existing {product-title} 4.9 cluster to this latest release, see xref:../updating/updating-cluster-cli.adoc[Updating a cluster using the CLI] for instructions.

> [[ocp-4-9-31]]

Reviewer: You have two brackets here [[ocp-4-9-31]] and it should be [id="ocp-4-9-31"] .

Reviewer: Can you fix 4.9.29 while you are at it see line 2760

Author: Got it! removed brackets for 4.9.31 and 4.9.29!

=== RHBA-2022:1605 - {product-title} 4.9.31 bug fix update

> Issued: 2022-05-02

Reviewer 2: I noticed that both of the advisories have the issue date 2022-05-03. You could potentially update this Release Notes entry to the 3rd May too.

{product-title} release 4.9.31, which includes security updates, is now available. The bug fixes that are included in the update are listed in the link:https://access.redhat.com/errata/RHBA-2022:1605[RHBA-2022:1605] advisory. The RPM packages that are included in the update are provided by the link:https://access.redhat.com/errata/RHBA-2022:1604[RHBA-2022:1604] advisory.

Space precluded documenting all of the container images for this release in the advisory. See the following article for notes on the container images in this release:

link:https://access.redhat.com/solutions/6955866[{product-title} 4.9.31 container image list]

[id="ocp-kubernetes-1-22-8-updates"]
==== Updates from Kubernetes 1.22.8

This update contains changes from Kubernetes 1.22.6 up to 1.22.8. More information can be found in the following changelogs: link:https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#v1226[1.22.6], link:https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#v1227[1.22.7], and link:https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md#v1228[1.22.8].

[id="ocp-4-9-31-bug-fixes"]
==== Bug fixes

> * Previously, a rewrite in the Application Programming Interface (API) prevented the creation of `DualStack` services without specifying the ipFamilyPolicy. This fix alerts users of the upcoming API change, ensuring awareness of the required ipFamilyPolicy specification beforehand. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2050632#[*BZ#2050632*])

Reviewer: ```suggestion
* Previously, a rewrite in the Application Programming Interface (API) prevented the creation of `DualStack` services without specifying the `ipFamilyPolicy`. This fix alerts users of the upcoming API change, ensuring awareness of the required `ipFamilyPolicy` specification beforehand. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2050632#[*BZ#2050632*])
```

Author: Updated! :)

> * Previously, a rewrite in the Application Programming Interface (API) prevented the creation of `DualStack` services without specifying the `ipFamilyPolicy`. This fix alerts users of the upcoming API change, ensuring awareness of the required `ipFamilyPolicy` specification beforehand. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2050632#[*BZ#2050632*])

Reviewer 2: You could change "Application Programming Interface" to lowercase "application programming interface". The only related reference I can find in the ISG is as follows:

> Do not use application program interface   noun
> Do not use. Use "application programming interface".

Reviewer 2: From a review of https://bugzilla.redhat.com/show_bug.cgi?id=2050632, my understanding is that from kube-apiserver 1.23 / OpenShift 4.10,  users explicitly specify either `ipFamilyPolicy: PreferDualStack` or `ipFamilyPolicy: RequireDualStack` for DualStack services to be valid. If I understand correct, regarding the OpenShift 4.9 context for these Release Notes, the change introduced is that there is now a warning stating that the API change in 4.10 is upcoming.

You could update this bug fix description to state this explicitly. This will prevent any ambiguity whereby readers might think that the API change applies to 4.9. For example:

> "In kube-apiserver 1.23 and {product-title} 4.10, users must specify either `ipFamilyPolicy: PreferDualStack` or `ipFamilyPolicy: RequireDualStack` for dual-stack services to be valid. Previously, no warning was provided in {product-title} 4.9 about the API requirement for {product-title} 4.10. This fix alerts users of the upcoming API change, ensuring awareness of the required `ipFamilyPolicy` specification beforehand."

> * Previously, no warning was provided in {product-title} 4.9 about API requiring users to specify either `ipFamilyPolicy: PreferDualStack` or `ipFamilyPolicy: RequireDualStack` for dual-stack services to be valid in {product-title} 4.10. This fix alerts users of the upcoming API change, ensuring awareness of the required specifications. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2050632#[*BZ#2050632*])

Reviewer 3: This is a little bit hard for me to track. And this seems like tough bug to describe. In particular, the use of passive voice is kind of throwing me. WDYT of something like this:
```suggestion
* Previously, the Kubernetes API server did not warn users about upcoming changes to the API for dual-stack services in {product-title} 4.10. This fix alerts users to specify either `ipFamilyPolicy: PreferDualStack` or `ipFamilyPolicy: RequireDualStack` for dual-stack services to be valid in {product-title} 4.10. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2050632#[*BZ#2050632*])
```

Author: Yes, this works well! I intended to make it as easily understood as possible. Thank you so much! I have made the change :)

> * Previously, the git import page failed to load due to performance improvement releases, causing the page to crash. With this fix, the git import page no longer fails to load and now runs as intended. (link:https://bugzilla.redhat.com/show_bug.cgi?id=2069621#[*BZ#2069621*])

Reviewer: Git is capitalized based on [SSG](https://redhat-documentation.github.io/supplementary-style-guide/#_g)

Reviewer: Since it is a page it should actually be 

> **Git import** page

[id="ocp-4-9-31-updating"]
==== Updating

> To upgrade an existing {product-title} 4.9 cluster to this latest release, see xref:../updating/updating-cluster-cli.adoc[Updating a cluster using the CLI] for instructions.

Reviewer 3: ```suggestion
To upgrade an existing {product-title} 4.9 cluster to this latest release, see xref:../updating/updating-cluster-cli.adoc#updating-cluster-cli[Updating a cluster using the CLI] for instructions.
```
I think you might have forgotten to add the context to the `xref` as Paul suggested. Would you please add it?

Author: The xref is stock text and will add updated as a continuous improvement. From what I understand from what Olivia told me, updating it here would mean all other z-streams will need to be updated.

Reviewer 2: Thank you for clarifying this as being handled holistically outside of this PR.
